refarch-integrations/refarch-s3-integration/refarch-s3-integration-core/src/main/java/de/muenchen/refarch/s3/integration/application/usecase/FileOperationsPresignedUrlUseCase.java

```diff
@@ -0,0 +1,193 @@
+package de.muenchen.refarch.s3.integration.application.usecase;
+
+import de.muenchen.refarch.s3.integration.adapter.in.rest.validation.FolderInFilePathValidator;
+import de.muenchen.refarch.s3.integration.application.port.in.FileOperationsPresignedUrlInPort;
+import de.muenchen.refarch.s3.integration.application.port.out.S3OutPort;
+import de.muenchen.refarch.s3.integration.domain.exception.FileExistenceException;
+import de.muenchen.refarch.s3.integration.domain.exception.FileSystemAccessException;
+import de.muenchen.refarch.s3.integration.domain.model.FileData;
+import de.muenchen.refarch.s3.integration.domain.model.PresignedUrl;
+import io.minio.http.Method;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class FileOperationsPresignedUrlUseCase implements FileOperationsPresignedUrlInPort {
+
+    private final S3OutPort s3OutPort;
+
+    /**
+     * Return the path to the folder for the given file path in the parameter.
+     * <p>
+     * pathToFile: FOLDER/SUBFOLDER/file.txt pathToFolder: FOLDER/SUBFOLDER
+     *
+     * @param pathToFile for which the path to folder should be returned.
+     * @return the path to the folder for the given path to file.
+     */
+    public static String getPathToFolder(final String pathToFile) {
+        return StringUtils.contains(pathToFile, FolderInFilePathValidator.SEPARATOR)
+                ? StringUtils.substringBeforeLast(pathToFile, FolderInFilePathValidator.SEPARATOR)
+                : StringUtils.EMPTY;
+    }
+
+    /**
+     * Get a list of presigned urls for all files in the paths. If the path is a file the presigned url
+     * for the file is returned.
+     *
+     * @param paths list of paths to files and/or folders
+     * @param action http method for the presigned url
+     * @param expiresInMinutes presigned url expiration time
+     * @return list of pre-signed urls.
+     * @throws FileSystemAccessException on S3 access errors.
+     * @throws FileExistenceException if file doesn't exist.
+     */
+    @Override
+    public List<PresignedUrl> getPresignedUrls(final List<String> paths, final Method action, final int expiresInMinutes)
+            throws FileSystemAccessException, FileExistenceException {
+        final List<PresignedUrl> presignedUrls = new ArrayList<>();
+        for (String p : paths) {
+            presignedUrls.addAll(this.getPresignedUrls(p, action, expiresInMinutes));
+        }
+        return presignedUrls;
+    }
```

Comment:
Could be refactored using functional programming (map)

Reply:
Think that is not working as `getPresignedUrls` can throw exception and i think that is not supported in lambda.

Reply:
Ok that's fine. We will see if they are reported by PMD later on.. ;)